torch/csrc/jit/codegen/cuda/test/test_gpu_transpose.cpp

```diff
@@ -710,6 +770,110 @@ TEST_F(NVFuserTest, FusionManualScheduleTransposeComplexDAG1_CUDA) {
       __FILE__);
 }
 
+// x->sin->transpose->cos->y
+TEST_F(NVFuserTest, FusionScheduleTransposeSmall_CUDA) {
+  Fusion fusion;
+  FusionGuard fg(&fusion);
+
+  auto tv0 = makeContigTensor(3);
+  fusion.addInput(tv0);
+  auto tv1 = sin(tv0);
+  auto tv2 = transpose(tv1, 1, 2);
+  auto tv3 = cos(tv2);
+  fusion.addOutput(tv3);
+
+  auto options = at::TensorOptions().dtype(at::kFloat).device(at::kCUDA, 0);
+  at::Tensor input = at::randn({1024, 2, 2}, options);
+
+  auto lparams = scheduleTranspose(&fusion, {input});
+
+  FusionExecutor fe;
+  fe.compileFusion(&fusion, {input}, lparams);
+  auto outputs = fe.runFusion({input}, lparams);
+
+  auto tv_ref = input.sin().transpose(1, 2).cos();
+
+  testValidate(&fusion, outputs, {input}, {tv_ref}, __LINE__, __FILE__);
+}
+
+// x->sin->transpose->cos->y
+TEST_F(NVFuserTest, FusionScheduleTransposeSmallInnerSize1_CUDA) {
+  Fusion fusion;
+  FusionGuard fg(&fusion);
+
+  auto tv0 = makeContigTensor(3);
+  fusion.addInput(tv0);
+  auto tv1 = sin(tv0);
+  auto tv2 = transpose(tv1, 1, 2);
+  auto tv3 = cos(tv2);
+  fusion.addOutput(tv3);
+
+  auto options = at::TensorOptions().dtype(at::kFloat).device(at::kCUDA, 0);
+  at::Tensor input = at::randn({64 * 1024 * 1024, 2, 2}, options);
+
+  auto lparams = scheduleTranspose(&fusion, {input});
+
+  FusionExecutor fe;
+  fe.compileFusion(&fusion, {input}, lparams);
+  auto outputs = fe.runFusion({input}, lparams);
+
+  auto tv_ref = input.sin().transpose(1, 2).cos();
+
+  testValidate(&fusion, outputs, {input}, {tv_ref}, __LINE__, __FILE__);
+}
+
+// x->sin->transpose->cos->y
+TEST_F(NVFuserTest, FusionScheduleTransposeSmallInnerSize2_CUDA) {
+  Fusion fusion;
+  FusionGuard fg(&fusion);
+
+  auto tv0 = makeContigTensor(3);
+  fusion.addInput(tv0);
+  auto tv1 = sin(tv0);
+  auto tv2 = transpose(tv1, 0, 2);
+  auto tv3 = cos(tv2);
+  fusion.addOutput(tv3);
+
+  auto options = at::TensorOptions().dtype(at::kFloat).device(at::kCUDA, 0);
+  at::Tensor input = at::randn({2, 64 * 1024 * 1024, 2}, options);
+
+  auto lparams = scheduleTranspose(&fusion, {input});
+
+  FusionExecutor fe;
+  fe.compileFusion(&fusion, {input}, lparams);
+  auto outputs = fe.runFusion({input}, lparams);
+
+  auto tv_ref = input.sin().transpose(0, 2).cos();
+
+  testValidate(&fusion, outputs, {input}, {tv_ref}, __LINE__, __FILE__);
+}
+
+// x->sin->transpose->cos->y
+TEST_F(NVFuserTest, FusionScheduleTransposeSmallInnerSize3_CUDA) {
```

Comment:
achieved: 154.589 GB/s on RTX 3090 (theoretical bandwidth: 936 GB/s)

Reply:
This is a compute-bound kernel...
![Screenshot_20220813_002533](https://user-images.githubusercontent.com/1032377/184473736-5547aaf2-2792-4780-826c-c8508d09e135.png)

Reply:
What does the code look like? Is it just a crazy amount of indexing math?

Reply:
Looks like it is just crazy indexing math:
```CUDA
__global__ void kernel1(Tensor<float, 8> T0, Tensor<float, 8> T3) {
  alignas(16) extern __shared__ char array[];
  unsigned smem_offset = 0;
  NVFUSER_DEFINE_MAGIC_ZERO
  smem_offset = alignBufferSize(smem_offset, 16);
  float* T4 = reinterpret_cast<float*>(array + smem_offset);
  smem_offset += ((((ceilDiv((ceilDiv((32 * 32), 2)), 128)) * 128) * 2) * sizeof(float));
  if (((((((((((((((((((((((((((((nvfuser_index_t)blockIdx.x) / (ceilDiv(((((4 * T0.size[1]) * T0.size[2]) * T0.size[3]) * T0.size[7]), 32))) * 32) + (((((((ceilDiv((ceilDiv((32 * 32), 2)), 128)) - 1) * 128) + ((nvfuser_index_t)threadIdx.x)) * 2) + 1) / 32)) / T0.size[4]) / T0.size[6]) / T0.size[5]) * 4) + (((((((((nvfuser_index_t)blockIdx.x) % (ceilDiv(((((4 * T0.size[1]) * T0.size[2]) * T0.size[3]) * T0.size[7]), 32))) * 32) + (((((((ceilDiv((ceilDiv((32 * 32), 2)), 128)) - 1) * 128) + ((nvfuser_index_t)threadIdx.x)) * 2) + 1) % 32)) / T0.size[7]) / T0.size[3]) / T0.size[2]) / T0.size[1])) < T0.size[0]) && ((((((((((nvfuser_index_t)blockIdx.x) % (ceilDiv(((((4 * T0.size[1]) * T0.size[2]) * T0.size[3]) * T0.size[7]), 32))) * 32) + (((((((ceilDiv((ceilDiv((32 * 32), 2)), 128)) - 1) * 128) + ((nvfuser_index_t)threadIdx.x)) * 2) + 1) % 32)) / T0.size[7]) / T0.size[3]) / T0.size[2]) % T0.size[1]) < T0.size[1])) && (((((((((nvfuser_index_t)blockIdx.x) % (ceilDiv(((((4 * T0.size[1]) * T0.size[2]) * T0.size[3]) * T0.size[7]), 32))) * 32) + (((((((ceilDiv((ceilDiv((32 * 32), 2)), 128)) - 1) * 128) + ((nvfuser_index_t)threadIdx.x)) * 2) + 1) % 32)) / T0.size[7]) / T0.size[3]) % T0.size[2]) < T0.size[2])) && ((((((((nvfuser_index_t)blockIdx.x) % (ceilDiv(((((4 * T0.size[1]) * T0.size[2]) * T0.size[3]) * T0.size[7]), 32))) * 32) + (((((((ceilDiv((ceilDiv((32 * 32), 2)), 128)) - 1) * 128) + ((nvfuser_index_t)threadIdx.x)) * 2) + 1) % 32)) / T0.size[7]) % T0.size[3]) < T0.size[3])) && (((((((nvfuser_index_t)blockIdx.x) / (ceilDiv(((((4 * T0.size[1]) * T0.size[2]) * T0.size[3]) * T0.size[7]), 32))) * 32) + (((((((ceilDiv((ceilDiv((32 * 32), 2)), 128)) - 1) * 128) + ((nvfuser_index_t)threadIdx.x)) * 2) + 1) / 32)) % T0.size[4]) < T0.size[4])) && (((((((((nvfuser_index_t)blockIdx.x) / (ceilDiv(((((4 * T0.size[1]) * T0.size[2]) * T0.size[3]) * T0.size[7]), 32))) * 32) + (((((((ceilDiv((ceilDiv((32 * 32), 2)), 128)) - 1) * 128) + ((nvfuser_index_t)threadIdx.x)) * 2) + 1) / 32)) / T0.size[4]) / T0.size[6]) % T0.size[5]) < T0.size[5])) && ((((((((nvfuser_index_t)blockIdx.x) / (ceilDiv(((((4 * T0.size[1]) * T0.size[2]) * T0.size[3]) * T0.size[7]), 32))) * 32) + (((((((ceilDiv((ceilDiv((32 * 32), 2)), 128)) - 1) * 128) + ((nvfuser_index_t)threadIdx.x)) * 2) + 1) / 32)) / T0.size[4]) % T0.size[6]) < T0.size[6])) && (((((((nvfuser_index_t)blockIdx.x) % (ceilDiv(((((4 * T0.size[1]) * T0.size[2]) * T0.size[3]) * T0.size[7]), 32))) * 32) + (((((((ceilDiv((ceilDiv((32 * 32), 2)), 128)) - 1) * 128) + ((nvfuser_index_t)threadIdx.x)) * 2) + 1) % 32)) % T0.size[7]) < T0.size[7])) && ((((((nvfuser_index_t)blockIdx.x) % (ceilDiv(((((4 * T0.size[1]) * T0.size[2]) * T0.size[3]) * T0.size[7]), 32))) * 32) + (((((((ceilDiv((ceilDiv((32 * 32), 2)), 128)) - 1) * 128) + ((nvfuser_index_t)threadIdx.x)) * 2) + 1) % 32)) < ((((4 * T0.size[1]) * T0.size[2]) * T0.size[3]) * T0.size[7]))) && (((((((((((nvfuser_index_t)blockIdx.x) / (ceilDiv(((((4 * T0.size[1]) * T0.size[2]) * T0.size[3]) * T0.size[7]), 32))) * 32) + (((((((ceilDiv((ceilDiv((32 * 32), 2)), 128)) - 1) * 128) + ((nvfuser_index_t)threadIdx.x)) * 2) + 1) % 32)) / T0.size[4]) / T0.size[6]) / T0.size[5]) * 4) + (((((((((nvfuser_index_t)blockIdx.x) % (ceilDiv(((((4 * T0.size[1]) * T0.size[2]) * T0.size[3]) * T0.size[7]), 32))) * 32) + (((((((ceilDiv((ceilDiv((32 * 32), 2)), 128)) - 1) * 128) + ((nvfuser_index_t)threadIdx.x)) * 2) + 1) / 32)) / T0.size[7]) / T0.size[3]) / T0.size[2]) / T0.size[1])) < T0.size[0])) && ((((((((((nvfuser_index_t)blockIdx.x) % (ceilDiv(((((4 * T0.size[1]) * T0.size[2]) * T0.size[3]) * T0.size[7]), 32))) * 32) + (((((((ceilDiv((ceilDiv((32 * 32), 2)), 128)) - 1) * 128) + ((nvfuser_index_t)threadIdx.x)) * 2) + 1) / 32)) / T0.size[7]) / T0.size[3]) / T0.size[2]) % T0.size[1]) < T0.size[1])) && (((((((((nvfuser_index_t)blockIdx.x) % (ceilDiv(((((4 * T0.size[1]) * T0.size[2]) * T0.size[3]) * T0.size[7]), 32))) * 32) + (((((((ceilDiv((ceilDiv((32 * 32), 2)), 128)) - 1) * 128) + ((nvfuser_index_t)threadIdx.x)) * 2) + 1) / 32)) / T0.size[7]) / T0.size[3]) % T0.size[2]) < T0.size[2])) && ((((((((nvfuser_index_t)blockIdx.x) % (ceilDiv(((((4 * T0.size[1]) * T0.size[2]) * T0.size[3]) * T0.size[7]), 32))) * 32) + (((((((ceilDiv((ceilDiv((32 * 32), 2)), 128)) - 1) * 128) + ((nvfuser_index_t)threadIdx.x)) * 2) + 1) / 32)) / T0.size[7]) % T0.size[3]) < T0.size[3])) && (((((((nvfuser_index_t)blockIdx.x) / (ceilDiv(((((4 * T0.size[1]) * T0.size[2]) * T0.size[3]) * T0.size[7]), 32))) * 32) + (((((((ceilDiv((ceilDiv((32 * 32), 2)), 128)) - 1) * 128) + ((nvfuser_index_t)threadIdx.x)) * 2) + 1) % 32)) % T0.size[4]) < T0.size[4])) && (((((((((nvfuser_index_t)blockIdx.x) / (ceilDiv(((((4 * T0.size[1]) * T0.size[2]) * T0.size[3]) * T0.size[7]), 32))) * 32) + (((((((ceilDiv((ceilDiv((32 * 32), 2)), 128)) - 1) * 128) + ((nvfuser_index_t)threadIdx.x)) * 2) + 1) % 32)) / T0.size[4]) / T0.size[6]) % T0.size[5]) < T0.size[5])) && ((((((((nvfuser_index_t)blockIdx.x) / (ceilDiv(((((4 * T0.size[1]) * T0.size[2]) * T0.size[3]) * T0.size[7]), 32))) * 32) + (((((((ceilDiv((ceilDiv((32 * 32), 2)), 128)) - 1) * 128) + ((nvfuser_index_t)threadIdx.x)) * 2) + 1) % 32)) / T0.size[4]) % T0.size[6]) < T0.size[6])) && (((((((nvfuser_index_t)blockIdx.x) % (ceilDiv(((((4 * T0.size[1]) * T0.size[2]) * T0.size[3]) * T0.size[7]), 32))) * 32) + (((((((ceilDiv((ceilDiv((32 * 32), 2)), 128)) - 1) * 128) + ((nvfuser_index_t)threadIdx.x)) * 2) + 1) / 32)) % T0.size[7]) < T0.size[7])) && ((((((nvfuser_index_t)blockIdx.x) % (ceilDiv(((((4 * T0.size[1]) * T0.size[2]) * T0.size[3]) * T0.size[7]), 32))) * 32) + (((((((ceilDiv((ceilDiv((32 * 32), 2)), 128)) - 1) * 128) + ((nvfuser_index_t)threadIdx.x)) * 2) + 1) / 32)) < ((((4 * T0.size[1]) * T0.size[2]) * T0.size[3]) * T0.size[7])))) {
    #pragma unroll
    for(nvfuser_index_t i154 = 0; i154 < (ceilDiv((ceilDiv((32 * 32), 2)), 128)); ++i154) {
      loadGeneric<float, 2>( &T4[(((((i154 * 128) + ((nvfuser_index_t)threadIdx.x)) * 2) / 32) * 32) + ((((i154 * 128) + ((nvfuser_index_t)threadIdx.x)) * 2) % 32)],  &T0[(((((((((((nvfuser_index_t)blockIdx.x) / (ceilDiv(((((4 * T0.size[1]) * T0.size[2]) * T0.size[3]) * T0.size[7]), 32))) * 32) + (((((i154 + nvfuser_zero) * 128) + ((nvfuser_index_t)threadIdx.x)) * 2) / 32)) / T0.size[4]) / T0.size[6]) / T0.size[5]) * 4) + (((((((((nvfuser_index_t)blockIdx.x) % (ceilDiv(((((4 * T0.size[1]) * T0.size[2]) * T0.size[3]) * T0.size[7]), 32))) * 32) + (((((i154 + nvfuser_zero) * 128) + ((nvfuser_index_t)threadIdx.x)) * 2) % 32)) / T0.size[7]) / T0.size[3]) / T0.size[2]) / T0.size[1])) * ((((((T0.size[7] * T0.size[6]) * T0.size[5]) * T0.size[4]) * T0.size[3]) * T0.size[2]) * T0.size[1])) + ((((((((((nvfuser_index_t)blockIdx.x) % (ceilDiv(((((4 * T0.size[1]) * T0.size[2]) * T0.size[3]) * T0.size[7]), 32))) * 32) + (((((i154 + nvfuser_zero) * 128) + ((nvfuser_index_t)threadIdx.x)) * 2) % 32)) / T0.size[7]) / T0.size[3]) / T0.size[2]) % T0.size[1]) * (((((T0.size[7] * T0.size[6]) * T0.size[5]) * T0.size[4]) * T0.size[3]) * T0.size[2])) + (((((((((nvfuser_index_t)blockIdx.x) % (ceilDiv(((((4 * T0.size[1]) * T0.size[2]) * T0.size[3]) * T0.size[7]), 32))) * 32) + (((((i154 + nvfuser_zero) * 128) + ((nvfuser_index_t)threadIdx.x)) * 2) % 32)) / T0.size[7]) / T0.size[3]) % T0.size[2]) * ((((T0.size[7] * T0.size[6]) * T0.size[5]) * T0.size[4]) * T0.size[3])) + ((((((((nvfuser_index_t)blockIdx.x) % (ceilDiv(((((4 * T0.size[1]) * T0.size[2]) * T0.size[3]) * T0.size[7]), 32))) * 32) + (((((i154 + nvfuser_zero) * 128) + ((nvfuser_index_t)threadIdx.x)) * 2) % 32)) / T0.size[7]) % T0.size[3]) * (((T0.size[7] * T0.size[6]) * T0.size[5]) * T0.size[4])) + (((((((nvfuser_index_t)blockIdx.x) / (ceilDiv(((((4 * T0.size[1]) * T0.size[2]) * T0.size[3]) * T0.size[7]), 32))) * 32) + (((((i154 + nvfuser_zero) * 128) + ((nvfuser_index_t)threadIdx.x)) * 2) / 32)) % T0.size[4]) * ((T0.size[7] * T0.size[6]) * T0.size[5])) + (((((((((nvfuser_index_t)blockIdx.x) / (ceilDiv(((((4 * T0.size[1]) * T0.size[2]) * T0.size[3]) * T0.size[7]), 32))) * 32) + (((((i154 + nvfuser_zero) * 128) + ((nvfuser_index_t)threadIdx.x)) * 2) / 32)) / T0.size[4]) / T0.size[6]) % T0.size[5]) * (T0.size[7] * T0.size[6])) + ((((((((nvfuser_index_t)blockIdx.x) / (ceilDiv(((((4 * T0.size[1]) * T0.size[2]) * T0.size[3]) * T0.size[7]), 32))) * 32) + (((((i154 + nvfuser_zero) * 128) + ((nvfuser_index_t)threadIdx.x)) * 2) / 32)) / T0.size[4]) % T0.size[6]) * T0.size[7]) + ((((((nvfuser_index_t)blockIdx.x) % (ceilDiv(((((4 * T0.size[1]) * T0.size[2]) * T0.size[3]) * T0.size[7]), 32))) * 32) + (((((i154 + nvfuser_zero) * 128) + ((nvfuser_index_t)threadIdx.x)) * 2) % 32)) % T0.size[7])]);
    }
    NVFUSER_UPDATE_MAGIC_ZERO
    Array<float, ((ceilDiv((ceilDiv((32 * 32), 2)), 128)) * 2), 2> T5;
    __barrier_sync(0);
    #pragma unroll
    for(nvfuser_index_t i156 = 0; i156 < (ceilDiv((ceilDiv((32 * 32), 2)), 128)); ++i156) {
      #pragma unroll
      for(nvfuser_index_t i157 = 0; i157 < 2; ++i157) {
        float T1[1];
        T1[0]
           = sinf(T4[((((((i156 * 128) + ((nvfuser_index_t)threadIdx.x)) * 2) + i157) % 32) * 32) + (((((i156 * 128) + ((nvfuser_index_t)threadIdx.x)) * 2) + i157) / 32)]);
        float T2[1];
        T2[0]
           = T1[0];
        T5[((i156 * 2) + i157)]
           = cosf(T2[0]);
      }
    }
    NVFUSER_UPDATE_MAGIC_ZERO
    #pragma unroll
    for(nvfuser_index_t i159 = 0; i159 < (ceilDiv((ceilDiv((32 * 32), 2)), 128)); ++i159) {
      loadLocalToGlobal<float, 2, false>( &T3[(((((((((((nvfuser_index_t)blockIdx.x) / (ceilDiv(((((4 * T0.size[1]) * T0.size[2]) * T0.size[3]) * T0.size[7]), 32))) * 32) + (((((i159 + nvfuser_zero) * 128) + ((nvfuser_index_t)threadIdx.x)) * 2) % 32)) / T0.size[4]) / T0.size[6]) / T0.size[5]) * 4) + (((((((((nvfuser_index_t)blockIdx.x) % (ceilDiv(((((4 * T0.size[1]) * T0.size[2]) * T0.size[3]) * T0.size[7]), 32))) * 32) + (((((i159 + nvfuser_zero) * 128) + ((nvfuser_index_t)threadIdx.x)) * 2) / 32)) / T0.size[7]) / T0.size[3]) / T0.size[2]) / T0.size[1])) * ((((((T0.size[4] * T0.size[6]) * T0.size[5]) * T0.size[7]) * T0.size[3]) * T0.size[2]) * T0.size[1])) + ((((((((((nvfuser_index_t)blockIdx.x) % (ceilDiv(((((4 * T0.size[1]) * T0.size[2]) * T0.size[3]) * T0.size[7]), 32))) * 32) + (((((i159 + nvfuser_zero) * 128) + ((nvfuser_index_t)threadIdx.x)) * 2) / 32)) / T0.size[7]) / T0.size[3]) / T0.size[2]) % T0.size[1]) * (((((T0.size[4] * T0.size[6]) * T0.size[5]) * T0.size[7]) * T0.size[3]) * T0.size[2])) + (((((((((nvfuser_index_t)blockIdx.x) % (ceilDiv(((((4 * T0.size[1]) * T0.size[2]) * T0.size[3]) * T0.size[7]), 32))) * 32) + (((((i159 + nvfuser_zero) * 128) + ((nvfuser_index_t)threadIdx.x)) * 2) / 32)) / T0.size[7]) / T0.size[3]) % T0.size[2]) * ((((T0.size[4] * T0.size[6]) * T0.size[5]) * T0.size[7]) * T0.size[3])) + ((((((((nvfuser_index_t)blockIdx.x) % (ceilDiv(((((4 * T0.size[1]) * T0.size[2]) * T0.size[3]) * T0.size[7]), 32))) * 32) + (((((i159 + nvfuser_zero) * 128) + ((nvfuser_index_t)threadIdx.x)) * 2) / 32)) / T0.size[7]) % T0.size[3]) * (((T0.size[4] * T0.size[6]) * T0.size[5]) * T0.size[7])) + (((((((nvfuser_index_t)blockIdx.x) % (ceilDiv(((((4 * T0.size[1]) * T0.size[2]) * T0.size[3]) * T0.size[7]), 32))) * 32) + (((((i159 + nvfuser_zero) * 128) + ((nvfuser_index_t)threadIdx.x)) * 2) / 32)) % T0.size[7]) * ((T0.size[4] * T0.size[6]) * T0.size[5])) + (((((((((nvfuser_index_t)blockIdx.x) / (ceilDiv(((((4 * T0.size[1]) * T0.size[2]) * T0.size[3]) * T0.size[7]), 32))) * 32) + (((((i159 + nvfuser_zero) * 128) + ((nvfuser_index_t)threadIdx.x)) * 2) % 32)) / T0.size[4]) / T0.size[6]) % T0.size[5]) * (T0.size[4] * T0.size[6])) + ((((((((nvfuser_index_t)blockIdx.x) / (ceilDiv(((((4 * T0.size[1]) * T0.size[2]) * T0.size[3]) * T0.size[7]), 32))) * 32) + (((((i159 + nvfuser_zero) * 128) + ((nvfuser_index_t)threadIdx.x)) * 2) % 32)) / T0.size[4]) % T0.size[6]) * T0.size[4]) + ((((((nvfuser_index_t)blockIdx.x) / (ceilDiv(((((4 * T0.size[1]) * T0.size[2]) * T0.size[3]) * T0.size[7]), 32))) * 32) + (((((i159 + nvfuser_zero) * 128) + ((nvfuser_index_t)threadIdx.x)) * 2) % 32)) % T0.size[4])], &T5[(i159 * 2)]);
    }
    NVFUSER_UPDATE_MAGIC_ZERO
  } else {
    #pragma unroll
    for(nvfuser_index_t i154 = 0; i154 < (ceilDiv((ceilDiv((32 * 32), 2)), 128)); ++i154) {
      int64_t i476;
      i476 = (((((((((nvfuser_index_t)blockIdx.x) / (ceilDiv(((((4 * T0.size[1]) * T0.size[2]) * T0.size[3]) * T0.size[7]), 32))) * 32) + (((((i154 + nvfuser_zero) * 128) + ((nvfuser_index_t)threadIdx.x)) * 2) / 32)) / T0.size[4]) / T0.size[6]) / T0.size[5]) * 4) + (((((((((nvfuser_index_t)blockIdx.x) % (ceilDiv(((((4 * T0.size[1]) * T0.size[2]) * T0.size[3]) * T0.size[7]), 32))) * 32) + (((((i154 + nvfuser_zero) * 128) + ((nvfuser_index_t)threadIdx.x)) * 2) % 32)) / T0.size[7]) / T0.size[3]) / T0.size[2]) / T0.size[1]);
      int64_t i460;
      i460 = ((((((nvfuser_index_t)blockIdx.x) % (ceilDiv(((((4 * T0.size[1]) * T0.size[2]) * T0.size[3]) * T0.size[7]), 32))) * 32) + (((((i154 + nvfuser_zero) * 128) + ((nvfuser_index_t)threadIdx.x)) * 2) % 32)) / T0.size[7]) % T0.size[3];
      int64_t i464;
      i464 = ((((((((nvfuser_index_t)blockIdx.x) % (ceilDiv(((((4 * T0.size[1]) * T0.size[2]) * T0.size[3]) * T0.size[7]), 32))) * 32) + (((((i154 + nvfuser_zero) * 128) + ((nvfuser_index_t)threadIdx.x)) * 2) % 32)) / T0.size[7]) / T0.size[3]) / T0.size[2]) % T0.size[1];
      int64_t i462;
      i462 = (((((((nvfuser_index_t)blockIdx.x) % (ceilDiv(((((4 * T0.size[1]) * T0.size[2]) * T0.size[3]) * T0.size[7]), 32))) * 32) + (((((i154 + nvfuser_zero) * 128) + ((nvfuser_index_t)threadIdx.x)) * 2) % 32)) / T0.size[7]) / T0.size[3]) % T0.size[2];
      int64_t i469;
      i469 = (((((nvfuser_index_t)blockIdx.x) / (ceilDiv(((((4 * T0.size[1]) * T0.size[2]) * T0.size[3]) * T0.size[7]), 32))) * 32) + (((((i154 + nvfuser_zero) * 128) + ((nvfuser_index_t)threadIdx.x)) * 2) / 32)) % T0.size[4];
      int64_t i473;
      i473 = (((((((nvfuser_index_t)blockIdx.x) / (ceilDiv(((((4 * T0.size[1]) * T0.size[2]) * T0.size[3]) * T0.size[7]), 32))) * 32) + (((((i154 + nvfuser_zero) * 128) + ((nvfuser_index_t)threadIdx.x)) * 2) / 32)) / T0.size[4]) / T0.size[6]) % T0.size[5];
      int64_t i458;
      i458 = (((((nvfuser_index_t)blockIdx.x) % (ceilDiv(((((4 * T0.size[1]) * T0.size[2]) * T0.size[3]) * T0.size[7]), 32))) * 32) + (((((i154 + nvfuser_zero) * 128) + ((nvfuser_index_t)threadIdx.x)) * 2) % 32)) % T0.size[7];
      int64_t i471;
      i471 = ((((((nvfuser_index_t)blockIdx.x) / (ceilDiv(((((4 * T0.size[1]) * T0.size[2]) * T0.size[3]) * T0.size[7]), 32))) * 32) + (((((i154 + nvfuser_zero) * 128) + ((nvfuser_index_t)threadIdx.x)) * 2) / 32)) / T0.size[4]) % T0.size[6];
      if ((((((((((((((((((((nvfuser_index_t)blockIdx.x) / (ceilDiv(((((4 * T0.size[1]) * T0.size[2]) * T0.size[3]) * T0.size[7]), 32))) * 32) + ((((((i154 + nvfuser_zero) * 128) + ((nvfuser_index_t)threadIdx.x)) * 2) + 1) / 32)) / T0.size[4]) / T0.size[6]) / T0.size[5]) * 4) + (((((((((nvfuser_index_t)blockIdx.x) % (ceilDiv(((((4 * T0.size[1]) * T0.size[2]) * T0.size[3]) * T0.size[7]), 32))) * 32) + ((((((i154 + nvfuser_zero) * 128) + ((nvfuser_index_t)threadIdx.x)) * 2) + 1) % 32)) / T0.size[7]) / T0.size[3]) / T0.size[2]) / T0.size[1])) < T0.size[0]) && ((((((((((nvfuser_index_t)blockIdx.x) % (ceilDiv(((((4 * T0.size[1]) * T0.size[2]) * T0.size[3]) * T0.size[7]), 32))) * 32) + ((((((i154 + nvfuser_zero) * 128) + ((nvfuser_index_t)threadIdx.x)) * 2) + 1) % 32)) / T0.size[7]) / T0.size[3]) / T0.size[2]) % T0.size[1]) < T0.size[1])) && (((((((((nvfuser_index_t)blockIdx.x) % (ceilDiv(((((4 * T0.size[1]) * T0.size[2]) * T0.size[3]) * T0.size[7]), 32))) * 32) + ((((((i154 + nvfuser_zero) * 128) + ((nvfuser_index_t)threadIdx.x)) * 2) + 1) % 32)) / T0.size[7]) / T0.size[3]) % T0.size[2]) < T0.size[2])) && ((((((((nvfuser_index_t)blockIdx.x) % (ceilDiv(((((4 * T0.size[1]) * T0.size[2]) * T0.size[3]) * T0.size[7]), 32))) * 32) + ((((((i154 + nvfuser_zero) * 128) + ((nvfuser_index_t)threadIdx.x)) * 2) + 1) % 32)) / T0.size[7]) % T0.size[3]) < T0.size[3])) && (((((((nvfuser_index_t)blockIdx.x) / (ceilDiv(((((4 * T0.size[1]) * T0.size[2]) * T0.size[3]) * T0.size[7]), 32))) * 32) + ((((((i154 + nvfuser_zero) * 128) + ((nvfuser_index_t)threadIdx.x)) * 2) + 1) / 32)) % T0.size[4]) < T0.size[4])) && (((((((((nvfuser_index_t)blockIdx.x) / (ceilDiv(((((4 * T0.size[1]) * T0.size[2]) * T0.size[3]) * T0.size[7]), 32))) * 32) + ((((((i154 + nvfuser_zero) * 128) + ((nvfuser_index_t)threadIdx.x)) * 2) + 1) / 32)) / T0.size[4]) / T0.size[6]) % T0.size[5]) < T0.size[5])) && ((((((((nvfuser_index_t)blockIdx.x) / (ceilDiv(((((4 * T0.size[1]) * T0.size[2]) * T0.size[3]) * T0.size[7]), 32))) * 32) + ((((((i154 + nvfuser_zero) * 128) + ((nvfuser_index_t)threadIdx.x)) * 2) + 1) / 32)) / T0.size[4]) % T0.size[6]) < T0.size[6])) && (((((((nvfuser_index_t)blockIdx.x) % (ceilDiv(((((4 * T0.size[1]) * T0.size[2]) * T0.size[3]) * T0.size[7]), 32))) * 32) + ((((((i154 + nvfuser_zero) * 128) + ((nvfuser_index_t)threadIdx.x)) * 2) + 1) % 32)) % T0.size[7]) < T0.size[7])) && ((((((nvfuser_index_t)blockIdx.x) % (ceilDiv(((((4 * T0.size[1]) * T0.size[2]) * T0.size[3]) * T0.size[7]), 32))) * 32) + ((((((i154 + nvfuser_zero) * 128) + ((nvfuser_index_t)threadIdx.x)) * 2) + 1) % 32)) < ((((4 * T0.size[1]) * T0.size[2]) * T0.size[3]) * T0.size[7])))) {
        loadGeneric<float, 2>( &T4[(((((i154 * 128) + ((nvfuser_index_t)threadIdx.x)) * 2) / 32) * 32) + ((((i154 * 128) + ((nvfuser_index_t)threadIdx.x)) * 2) % 32)],  &T0[(i476 * ((((((T0.size[7] * T0.size[6]) * T0.size[5]) * T0.size[4]) * T0.size[3]) * T0.size[2]) * T0.size[1])) + (i464 * (((((T0.size[7] * T0.size[6]) * T0.size[5]) * T0.size[4]) * T0.size[3]) * T0.size[2])) + (i462 * ((((T0.size[7] * T0.size[6]) * T0.size[5]) * T0.size[4]) * T0.size[3])) + (i460 * (((T0.size[7] * T0.size[6]) * T0.size[5]) * T0.size[4])) + (i469 * ((T0.size[7] * T0.size[6]) * T0.size[5])) + (i473 * (T0.size[7] * T0.size[6])) + (i471 * T0.size[7]) + i458]);
      }
    }
    NVFUSER_UPDATE_MAGIC_ZERO
    Array<float, ((ceilDiv((ceilDiv((32 * 32), 2)), 128)) * 2), 2> T5;
    __barrier_sync(0);
    #pragma unroll
    for(nvfuser_index_t i156 = 0; i156 < (ceilDiv((ceilDiv((32 * 32), 2)), 128)); ++i156) {
      #pragma unroll
      for(nvfuser_index_t i157 = 0; i157 < 2; ++i157) {
        int64_t i2269;
        i2269 = ((((nvfuser_index_t)blockIdx.x) % (ceilDiv(((((4 * T0.size[1]) * T0.size[2]) * T0.size[3]) * T0.size[7]), 32))) * 32) + (((((i156 * 128) + ((nvfuser_index_t)threadIdx.x)) * 2) + (i157 + nvfuser_zero)) / 32);
        int64_t i2109;
        i2109 = (((((((((nvfuser_index_t)blockIdx.x) / (ceilDiv(((((4 * T0.size[1]) * T0.size[2]) * T0.size[3]) * T0.size[7]), 32))) * 32) + (((((i156 * 128) + ((nvfuser_index_t)threadIdx.x)) * 2) + (i157 + nvfuser_zero)) % 32)) / T0.size[4]) / T0.size[6]) / T0.size[5]) * 4) + (((((((((nvfuser_index_t)blockIdx.x) % (ceilDiv(((((4 * T0.size[1]) * T0.size[2]) * T0.size[3]) * T0.size[7]), 32))) * 32) + (((((i156 * 128) + ((nvfuser_index_t)threadIdx.x)) * 2) + (i157 + nvfuser_zero)) / 32)) / T0.size[7]) / T0.size[3]) / T0.size[2]) / T0.size[1]);
        int64_t i2172;
        i2172 = ((((((nvfuser_index_t)blockIdx.x) % (ceilDiv(((((4 * T0.size[1]) * T0.size[2]) * T0.size[3]) * T0.size[7]), 32))) * 32) + (((((i156 * 128) + ((nvfuser_index_t)threadIdx.x)) * 2) + (i157 + nvfuser_zero)) / 32)) / T0.size[7]) % T0.size[3];
        int64_t i2131;
        i2131 = ((((((((nvfuser_index_t)blockIdx.x) % (ceilDiv(((((4 * T0.size[1]) * T0.size[2]) * T0.size[3]) * T0.size[7]), 32))) * 32) + (((((i156 * 128) + ((nvfuser_index_t)threadIdx.x)) * 2) + (i157 + nvfuser_zero)) / 32)) / T0.size[7]) / T0.size[3]) / T0.size[2]) % T0.size[1];
        int64_t i2152;
        i2152 = (((((((nvfuser_index_t)blockIdx.x) % (ceilDiv(((((4 * T0.size[1]) * T0.size[2]) * T0.size[3]) * T0.size[7]), 32))) * 32) + (((((i156 * 128) + ((nvfuser_index_t)threadIdx.x)) * 2) + (i157 + nvfuser_zero)) / 32)) / T0.size[7]) / T0.size[3]) % T0.size[2];
        int64_t i2191;
        i2191 = (((((nvfuser_index_t)blockIdx.x) / (ceilDiv(((((4 * T0.size[1]) * T0.size[2]) * T0.size[3]) * T0.size[7]), 32))) * 32) + (((((i156 * 128) + ((nvfuser_index_t)threadIdx.x)) * 2) + (i157 + nvfuser_zero)) % 32)) % T0.size[4];
        int64_t i2212;
        i2212 = (((((((nvfuser_index_t)blockIdx.x) / (ceilDiv(((((4 * T0.size[1]) * T0.size[2]) * T0.size[3]) * T0.size[7]), 32))) * 32) + (((((i156 * 128) + ((nvfuser_index_t)threadIdx.x)) * 2) + (i157 + nvfuser_zero)) % 32)) / T0.size[4]) / T0.size[6]) % T0.size[5];
        int64_t i2251;
        i2251 = (((((nvfuser_index_t)blockIdx.x) % (ceilDiv(((((4 * T0.size[1]) * T0.size[2]) * T0.size[3]) * T0.size[7]), 32))) * 32) + (((((i156 * 128) + ((nvfuser_index_t)threadIdx.x)) * 2) + (i157 + nvfuser_zero)) / 32)) % T0.size[7];
        int64_t i2232;
        i2232 = ((((((nvfuser_index_t)blockIdx.x) / (ceilDiv(((((4 * T0.size[1]) * T0.size[2]) * T0.size[3]) * T0.size[7]), 32))) * 32) + (((((i156 * 128) + ((nvfuser_index_t)threadIdx.x)) * 2) + (i157 + nvfuser_zero)) % 32)) / T0.size[4]) % T0.size[6];
        float T1[1];
        if ((((((((((i2109 < T0.size[0]) && (i2131 < T0.size[1])) && (i2152 < T0.size[2])) && (i2172 < T0.size[3])) && (i2191 < T0.size[4])) && (i2212 < T0.size[5])) && (i2232 < T0.size[6])) && (i2251 < T0.size[7])) && (i2269 < ((((4 * T0.size[1]) * T0.size[2]) * T0.size[3]) * T0.size[7])))) {
          T1[0]
             = sinf(T4[((((((i156 * 128) + ((nvfuser_index_t)threadIdx.x)) * 2) + i157) % 32) * 32) + (((((i156 * 128) + ((nvfuser_index_t)threadIdx.x)) * 2) + i157) / 32)]);
        }
        float T2[1];
        if ((((((((((i2109 < T0.size[0]) && (i2131 < T0.size[1])) && (i2152 < T0.size[2])) && (i2172 < T0.size[3])) && (i2251 < T0.size[7])) && (i2212 < T0.size[5])) && (i2232 < T0.size[6])) && (i2191 < T0.size[4])) && (i2269 < ((((4 * T0.size[1]) * T0.size[2]) * T0.size[3]) * T0.size[7])))) {
          T2[0]
             = T1[0];
        }
        if ((((((((((i2109 < T0.size[0]) && (i2131 < T0.size[1])) && (i2152 < T0.size[2])) && (i2172 < T0.size[3])) && (i2251 < T0.size[7])) && (i2212 < T0.size[5])) && (i2232 < T0.size[6])) && (i2191 < T0.size[4])) && (i2269 < ((((4 * T0.size[1]) * T0.size[2]) * T0.size[3]) * T0.size[7])))) {
          T5[((i156 * 2) + i157)]
             = cosf(T2[0]);
        }
      }
    }
    NVFUSER_UPDATE_MAGIC_ZERO
    #pragma unroll
    for(nvfuser_index_t i159 = 0; i159 < (ceilDiv((ceilDiv((32 * 32), 2)), 128)); ++i159) {
      int64_t i642;
      i642 = (((((((((nvfuser_index_t)blockIdx.x) / (ceilDiv(((((4 * T0.size[1]) * T0.size[2]) * T0.size[3]) * T0.size[7]), 32))) * 32) + (((((i159 + nvfuser_zero) * 128) + ((nvfuser_index_t)threadIdx.x)) * 2) % 32)) / T0.size[4]) / T0.size[6]) / T0.size[5]) * 4) + (((((((((nvfuser_index_t)blockIdx.x) % (ceilDiv(((((4 * T0.size[1]) * T0.size[2]) * T0.size[3]) * T0.size[7]), 32))) * 32) + (((((i159 + nvfuser_zero) * 128) + ((nvfuser_index_t)threadIdx.x)) * 2) / 32)) / T0.size[7]) / T0.size[3]) / T0.size[2]) / T0.size[1]);
      int64_t i626;
      i626 = ((((((nvfuser_index_t)blockIdx.x) % (ceilDiv(((((4 * T0.size[1]) * T0.size[2]) * T0.size[3]) * T0.size[7]), 32))) * 32) + (((((i159 + nvfuser_zero) * 128) + ((nvfuser_index_t)threadIdx.x)) * 2) / 32)) / T0.size[7]) % T0.size[3];
      int64_t i630;
      i630 = ((((((((nvfuser_index_t)blockIdx.x) % (ceilDiv(((((4 * T0.size[1]) * T0.size[2]) * T0.size[3]) * T0.size[7]), 32))) * 32) + (((((i159 + nvfuser_zero) * 128) + ((nvfuser_index_t)threadIdx.x)) * 2) / 32)) / T0.size[7]) / T0.size[3]) / T0.size[2]) % T0.size[1];
      int64_t i628;
      i628 = (((((((nvfuser_index_t)blockIdx.x) % (ceilDiv(((((4 * T0.size[1]) * T0.size[2]) * T0.size[3]) * T0.size[7]), 32))) * 32) + (((((i159 + nvfuser_zero) * 128) + ((nvfuser_index_t)threadIdx.x)) * 2) / 32)) / T0.size[7]) / T0.size[3]) % T0.size[2];
      int64_t i635;
      i635 = (((((nvfuser_index_t)blockIdx.x) / (ceilDiv(((((4 * T0.size[1]) * T0.size[2]) * T0.size[3]) * T0.size[7]), 32))) * 32) + (((((i159 + nvfuser_zero) * 128) + ((nvfuser_index_t)threadIdx.x)) * 2) % 32)) % T0.size[4];
      int64_t i639;
      i639 = (((((((nvfuser_index_t)blockIdx.x) / (ceilDiv(((((4 * T0.size[1]) * T0.size[2]) * T0.size[3]) * T0.size[7]), 32))) * 32) + (((((i159 + nvfuser_zero) * 128) + ((nvfuser_index_t)threadIdx.x)) * 2) % 32)) / T0.size[4]) / T0.size[6]) % T0.size[5];
      int64_t i624;
      i624 = (((((nvfuser_index_t)blockIdx.x) % (ceilDiv(((((4 * T0.size[1]) * T0.size[2]) * T0.size[3]) * T0.size[7]), 32))) * 32) + (((((i159 + nvfuser_zero) * 128) + ((nvfuser_index_t)threadIdx.x)) * 2) / 32)) % T0.size[7];
      int64_t i637;
      i637 = ((((((nvfuser_index_t)blockIdx.x) / (ceilDiv(((((4 * T0.size[1]) * T0.size[2]) * T0.size[3]) * T0.size[7]), 32))) * 32) + (((((i159 + nvfuser_zero) * 128) + ((nvfuser_index_t)threadIdx.x)) * 2) % 32)) / T0.size[4]) % T0.size[6];
      if ((((((((((((((((((((nvfuser_index_t)blockIdx.x) / (ceilDiv(((((4 * T0.size[1]) * T0.size[2]) * T0.size[3]) * T0.size[7]), 32))) * 32) + ((((((i159 + nvfuser_zero) * 128) + ((nvfuser_index_t)threadIdx.x)) * 2) + 1) % 32)) / T0.size[4]) / T0.size[6]) / T0.size[5]) * 4) + (((((((((nvfuser_index_t)blockIdx.x) % (ceilDiv(((((4 * T0.size[1]) * T0.size[2]) * T0.size[3]) * T0.size[7]), 32))) * 32) + ((((((i159 + nvfuser_zero) * 128) + ((nvfuser_index_t)threadIdx.x)) * 2) + 1) / 32)) / T0.size[7]) / T0.size[3]) / T0.size[2]) / T0.size[1])) < T0.size[0]) && ((((((((((nvfuser_index_t)blockIdx.x) % (ceilDiv(((((4 * T0.size[1]) * T0.size[2]) * T0.size[3]) * T0.size[7]), 32))) * 32) + ((((((i159 + nvfuser_zero) * 128) + ((nvfuser_index_t)threadIdx.x)) * 2) + 1) / 32)) / T0.size[7]) / T0.size[3]) / T0.size[2]) % T0.size[1]) < T0.size[1])) && (((((((((nvfuser_index_t)blockIdx.x) % (ceilDiv(((((4 * T0.size[1]) * T0.size[2]) * T0.size[3]) * T0.size[7]), 32))) * 32) + ((((((i159 + nvfuser_zero) * 128) + ((nvfuser_index_t)threadIdx.x)) * 2) + 1) / 32)) / T0.size[7]) / T0.size[3]) % T0.size[2]) < T0.size[2])) && ((((((((nvfuser_index_t)blockIdx.x) % (ceilDiv(((((4 * T0.size[1]) * T0.size[2]) * T0.size[3]) * T0.size[7]), 32))) * 32) + ((((((i159 + nvfuser_zero) * 128) + ((nvfuser_index_t)threadIdx.x)) * 2) + 1) / 32)) / T0.size[7]) % T0.size[3]) < T0.size[3])) && (((((((nvfuser_index_t)blockIdx.x) % (ceilDiv(((((4 * T0.size[1]) * T0.size[2]) * T0.size[3]) * T0.size[7]), 32))) * 32) + ((((((i159 + nvfuser_zero) * 128) + ((nvfuser_index_t)threadIdx.x)) * 2) + 1) / 32)) % T0.size[7]) < T0.size[7])) && (((((((((nvfuser_index_t)blockIdx.x) / (ceilDiv(((((4 * T0.size[1]) * T0.size[2]) * T0.size[3]) * T0.size[7]), 32))) * 32) + ((((((i159 + nvfuser_zero) * 128) + ((nvfuser_index_t)threadIdx.x)) * 2) + 1) % 32)) / T0.size[4]) / T0.size[6]) % T0.size[5]) < T0.size[5])) && ((((((((nvfuser_index_t)blockIdx.x) / (ceilDiv(((((4 * T0.size[1]) * T0.size[2]) * T0.size[3]) * T0.size[7]), 32))) * 32) + ((((((i159 + nvfuser_zero) * 128) + ((nvfuser_index_t)threadIdx.x)) * 2) + 1) % 32)) / T0.size[4]) % T0.size[6]) < T0.size[6])) && (((((((nvfuser_index_t)blockIdx.x) / (ceilDiv(((((4 * T0.size[1]) * T0.size[2]) * T0.size[3]) * T0.size[7]), 32))) * 32) + ((((((i159 + nvfuser_zero) * 128) + ((nvfuser_index_t)threadIdx.x)) * 2) + 1) % 32)) % T0.size[4]) < T0.size[4])) && ((((((nvfuser_index_t)blockIdx.x) % (ceilDiv(((((4 * T0.size[1]) * T0.size[2]) * T0.size[3]) * T0.size[7]), 32))) * 32) + ((((((i159 + nvfuser_zero) * 128) + ((nvfuser_index_t)threadIdx.x)) * 2) + 1) / 32)) < ((((4 * T0.size[1]) * T0.size[2]) * T0.size[3]) * T0.size[7])))) {
        loadLocalToGlobal<float, 2, false>( &T3[(i642 * ((((((T0.size[4] * T0.size[6]) * T0.size[5]) * T0.size[7]) * T0.size[3]) * T0.size[2]) * T0.size[1])) + (i630 * (((((T0.size[4] * T0.size[6]) * T0.size[5]) * T0.size[7]) * T0.size[3]) * T0.size[2])) + (i628 * ((((T0.size[4] * T0.size[6]) * T0.size[5]) * T0.size[7]) * T0.size[3])) + (i626 * (((T0.size[4] * T0.size[6]) * T0.size[5]) * T0.size[7])) + (i624 * ((T0.size[4] * T0.size[6]) * T0.size[5])) + (i639 * (T0.size[4] * T0.size[6])) + (i637 * T0.size[4]) + i635], &T5[(i159 * 2)]);
      }
    }
    NVFUSER_UPDATE_MAGIC_ZERO
  }
}

```

Reply:
Wow.

Reply:
We should check to see if contig merge's aren't helping here. There used to be a condition where we needed to merge inside to outside, but I thought that condition wasn't required anymore. For now we can move on, but would be interesting to revisit at some point.